modules/social_features/social_core/src/Plugin/Field/FieldType/ExternalIdentifierItem.php

```diff
@@ -0,0 +1,221 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\social_core\Plugin\Field\FieldType;
+
+use Drupal\Core\Entity\EntityInterface;
+use Drupal\Core\Field\FieldItemBase;
+use Drupal\Core\Field\FieldStorageDefinitionInterface;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\StringTranslation\TranslatableMarkup;
+use Drupal\Core\TypedData\DataDefinition;
+use Drupal\Core\TypedData\DataReferenceTargetDefinition;
+
+/**
+ * Defines the 'social_external_identifier' field type.
+ *
+ * Supported settings (below the definition's 'settings' key) are:
+ * - target_type: The entity type to reference. Required.
+ *
+ * @FieldType(
+ *   id = "social_external_identifier",
+ *   label = @Translation("External Identifier"),
+ *   description = @Translation("Define external Identifier and external Owner/Consumer."),
+ *   default_widget = "social_external_identifier_default_widget",
+ *   default_formatter = "social_external_identifier_default_formatter",
+ *   constraints = {
+ *     "ExternalIdentifierEmptySubfieldsConstraint" = {},
+ *     "ExternalIdentifierExternalOwnerTargetTypeConstraint" = {},
+ *     "ExternalIdentifierExternalOwnerIdConstraint" = {},
+ *     "ComplexData" = {
+ *       "external_id" = {
+ *         "Length" = {"max" = 225}
+ *        }
+ *      }
+ *   }
+ * )
+ */
+final class ExternalIdentifierItem extends FieldItemBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isEmpty(): bool {
+    // If at least one of 3 subfields is not empty, field is not empty.
+    // Subfields are:
+    // - external_id,
+    // - external_owner_target_type
+    // - external_owner_id
+    //
+    // Note, even that external_owner_id is int, we are checking is as string
+    // because of some Drupal bugs that int values ane not matching type hints.
+    // See: https://www.drupal.org/project/drupal/issues/3441689
+    // See: https://www.drupal.org/project/drupal/issues/3224376
+    //
+    // Also note that the ExternalIdentifierEmptySubfieldsConstraintValidator
+    // prevents the field from being saved if all subfields do not have set
+    // values.
+    foreach ($this->getValue() as $value) {
+      if ($value !== '' && $value !== NULL) {
+        return FALSE;
+      }
+    }
+
+    return TRUE;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function propertyDefinitions(FieldStorageDefinitionInterface $field_definition): array {
+    $properties['external_id'] = DataDefinition::create('string')
+      ->setLabel(t('External ID'))
+      ->setRequired(TRUE);
+    $properties['external_owner_target_type'] = DataReferenceTargetDefinition::create('string')
+      ->setLabel(t('Target Entity Type'))
+      ->setRequired(TRUE);
+    // @todo See if we can improve this subfield to use
+    //   DataReferenceDefinition or DataDynamicReferenceDefinition.
```

Comment:
short answer, we might want to improve this in future,
long answer, decision records https://open-social.slite.com/app/docs/pA97W-b4Z7PkiK

Reply:
I think a `@todo` should only survive a merge if it's linked to some Jira ticket so that we have an idea of what needs to happen and there's a non-zero chance it'll get picked up. The Jira issue is also a good place to point to the ADR and describe what/why we need this.

If we have a `@todo` in the code that's not linked to something outside of our code-base then it's likely to just sit there forever and only puzzle archeologists in the future :)

Reply:
Ok based on your comment @Kingdutch I will remove this todo and if we want to improve this we still have ADR where everything is documented.